File: 1691-maximum-height-by-stacking-cuboids/1691-maximum-height-by-stacking-cuboids.cpp
class Solution {
public:
    int f(vector<vector<int>>& dp , vector<vector<int>>& cuboids , int i , int prev){
        int n = cuboids.size();
        if(i  >= n) return 0;
        if(dp[i][prev+1] != -1) return dp[i][prev+1];
        int w1 = cuboids[i][0] ,  l1 = cuboids[i][1] , h1 = cuboids[i][2];
        int take = 0;
        if(prev == -1 || (cuboids[prev][0] <= w1 && cuboids[prev][1] <= l1 && cuboids[prev][2] <= h1)){
            take = h1 + f(dp , cuboids , i + 1 , i);
        }
        int notTake = f(dp , cuboids , i + 1 , prev);
        return dp[i][prev+1] = max(take , notTake); 
    }
    int maxHeight(vector<vector<int>>& cuboids) {
        int n = cuboids.size();
        for(auto& e : cuboids){
            sort(begin(e) , end(e));
        }
        sort(cuboids.begin() , cuboids.end()); 
        vector<vector<int>>dp(n + 1 , vector<int>(n + 1 , -1));
        return f(dp , cuboids , 0 , -1);
    }
};

File: 1691-maximum-height-by-stacking-cuboids/1691-maximum-height-by-stacking-cuboids.java
class Solution {
    int f(int[][] dp , int[][] cuboids , int i , int prev){
        int n = cuboids.length;
        if(i >= n) return 0;
        if(dp[i][prev + 1] != -1) return dp[i][prev + 1];
        int w = cuboids[i][0] , l = cuboids[i][1] , h = cuboids[i][2];
        int take = 0;
        if(prev == -1 || cuboids[prev][0] <= w && cuboids[prev][1] <= l && cuboids[prev][2] <= h){
            take = h + f(dp , cuboids , i + 1 , i);
        }
        int notTake = f(dp , cuboids , i + 1 , prev);
        return dp[i][prev + 1] = Math.max(take , notTake);
    }
    public int maxHeight(int[][] cuboids) {
        int n = cuboids.length;
        int[][] dp = new int[n + 1][n + 1];
        for(int[] row : dp){
            Arrays.fill(row , -1);
        }
        for(int[] row : cuboids){
            Arrays.sort(row);
        }
        Arrays.sort(cuboids , (a , b)->{
            if(a[0] != b[0]) return Integer.compare(a[0] , b[0]);
            if(a[1] != b[1]) return Integer.compare(a[1] , b[1]);
            return Integer.compare(a[2] , b[2]);
        });
        return f(dp , cuboids , 0 , -1);
    }
}